pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java

```diff
@@ -75,4 +95,72 @@ public WorkflowResponse buildStateResponse(Handle handle, String userGuid, Workf
+            String userGuid,
+            long studyId
+    ) {
+        if (nextState.getType() != StateType.ACTIVITY) {
+            return;
+        }
+        ActivityState activityState = (ActivityState) nextState;
+        ActivityDto activityDto = handle.attach(JdbiActivity.class).queryActivityById(activityState.getActivityId());
+        if (activityDto.getMaxInstancesPerUser() != null && activityDto.getMaxInstancesPerUser() == 0) {
+            return;
+        }
+        String instanceGuid = handle.attach(JdbiActivityInstance.class)
+                .findLatestInstanceGuidByUserGuidAndActivityId(userGuid, activityState.getActivityId())
+                .orElse(null);
+        boolean activityInstanceIsMissing = instanceGuid == null;
+        if (activityInstanceIsMissing) {
+            ActivityInstanceDto instanceDto = handle.attach(ActivityInstanceDao.class)
+                    .insertInstance(activityState.getActivityId(), operatorGuid, userGuid, InstanceStatusType.CREATED, false);
+            LOG.info("Created start activity instance with guid '{}' for user guid '{}' using operator guid '{}' and activity id {}",
+                    instanceDto.getGuid(), userGuid, operatorGuid, activityState.getActivityId());
+            processActionsForActivityCreationSignal(
+                    handle,
+                    operatorGuid,
+                    instanceDto.getParticipantId(),
+                    userGuid,
+                    instanceDto.getId(),
+                    activityState.getActivityId(),
+                    studyId
+            );
+            LOG.info(
+                    "Processed actions for the activity instance {} of activity {} triggered by the creation signal",
+                    instanceDto.getGuid(),
+                    activityState.getActivityId()
+            );
+        } else {
+            LOG.info("User guid '{}' already has activity instance with guid '{}', nothing to create", userGuid, instanceGuid);
+        }
+    }
+
+    private void processActionsForActivityCreationSignal(
+            Handle handle,
+            String operatorGuid,
+            long participantId,
+            String participantGuid,
+            long instanceId,
+            long activityId,
+            long studyId
+    ) {
+        // The DAO method returns optional while the column in "study_activity.study_id" is NOT NULL
+        // This means that "Optional<Long>" can be replaced with just "long"
+        long operatorId = handle.attach(JdbiUser.class).getUserIdByGuid(operatorGuid);
+        EventSignal activityCreationSignal = new ActivityInstanceStatusChangeSignal(
+                operatorId,
+                participantId,
+                participantGuid,
+                instanceId,
+                activityId,
+                studyId,
+                InstanceStatusType.CREATED
+        );
+        EventService.getInstance().processAllActionsForEventSignal(handle, activityCreationSignal);
+    }
 }
```

Comment:
Ok, so that's how the post-instance-creation event processing should look like according to my understanding: we publish an activity instance creation signal and then ask the `EventService` to process all the event configurations associated with it. @yufengwng please point out if I'm missing anything here.

Reply:
Looks right according to my understanding as well.